File: project-{topic}-complete/app/services/task_service.py
```python
import logging
from app.extensions import db
from app.models.task import Task
from app.models.project import Project
from app.models.user import User
from app.models.comment import Comment
from app.utils.decorators import log_service_operation
from app.utils.exceptions import ResourceNotFound, InvalidInput

logger = logging.getLogger(__name__)

class TaskService:
    @staticmethod
    @log_service_operation
    def create_task(data):
        title = data.get('title')
        project_id = data.get('project_id')
        creator_id = data.get('creator_id')

        if not title or not project_id or not creator_id:
            raise InvalidInput("Title, project ID, and creator ID are required for a task.")

        if not Project.query.get(project_id):
            raise ResourceNotFound(f"Project with ID '{project_id}' not found.")
        if not User.query.get(creator_id):
            raise ResourceNotFound(f"Creator with ID '{creator_id}' not found.")
        if data.get('assigned_to_id') and not User.query.get(data['assigned_to_id']):
            raise ResourceNotFound(f"Assignee with ID '{data['assigned_to_id']}' not found.")

        try:
            task = Task(
                title=title,
                description=data.get('description'),
                project_id=project_id,
                creator_id=creator_id,
                assigned_to_id=data.get('assigned_to_id'),
                status=data.get('status', 'open'),
                priority=data.get('priority', 'medium'),
                due_date=data.get('due_date')
            )
            db.session.add(task)
            db.session.commit()
            logger.info(f"Task created: {task.title}")
            return task
        except Exception as e:
            db.session.rollback()
            logger.error(f"Error creating task: {e}")
            raise

    @staticmethod
    @log_service_operation
    def get_all_tasks(page=1, per_page=10, project_id=None, assigned_to_id=None, status=None, priority=None):
        query = Task.query
        if project_id:
            query = query.filter_by(project_id=project_id)
        if assigned_to_id:
            query = query.filter_by(assigned_to_id=assigned_to_id)
        if status:
            query = query.filter_by(status=status)
        if priority:
            query = query.filter_by(priority=priority)

        tasks = query.paginate(page=page, per_page=per_page, error_out=False)
        return tasks

    @staticmethod
    @log_service_operation
    def get_task_by_id(task_id):
        task = Task.query.get(task_id)
        if not task:
            raise ResourceNotFound(f"Task with ID '{task_id}' not found.")
        return task

    @staticmethod
    @log_service_operation
    def update_task(task_id, data):
        task = TaskService.get_task_by_id(task_id)

        if 'project_id' in data and not Project.query.get(data['project_id']):
            raise ResourceNotFound(f"New project with ID '{data['project_id']}' not found.")
        if 'assigned_to_id' in data and data['assigned_to_id'] and not User.query.get(data['assigned_to_id']):
            raise ResourceNotFound(f"New assignee with ID '{data['assigned_to_id']}' not found.")
        if 'creator_id' in data and not User.query.get(data['creator_id']):
            raise ResourceNotFound(f"New creator with ID '{data['creator_id']}' not found.")

        try:
            task.title = data.get('title', task.title)
            task.description = data.get('description', task.description)
            task.project_id = data.get('project_id', task.project_id)
            task.creator_id = data.get('creator_id', task.creator_id)
            task.assigned_to_id = data.get('assigned_to_id', task.assigned_to_id)
            task.status = data.get('status', task.status)
            task.priority = data.get('priority', task.priority)
            task.due_date = data.get('due_date', task.due_date)
            db.session.commit()
            logger.info(f"Task updated: {task.title}")
            return task
        except Exception as e:
            db.session.rollback()
            logger.error(f"Error updating task {task_id}: {e}")
            raise

    @staticmethod
    @log_service_operation
    def delete_task(task_id):
        task = TaskService.get_task_by_id(task_id)
        try:
            db.session.delete(task)
            db.session.commit()
            logger.info(f"Task deleted: {task_id}")
            return True
        except Exception as e:
            db.session.rollback()
            logger.error(f"Error deleting task {task_id}: {e}")
            raise

    @staticmethod
    @log_service_operation
    def add_comment_to_task(task_id, content, author_id):
        task = TaskService.get_task_by_id(task_id)
        if not User.query.get(author_id):
            raise ResourceNotFound(f"Author with ID '{author_id}' not found.")
        
        try:
            comment = Comment(content=content, task_id=task_id, author_id=author_id)
            db.session.add(comment)
            db.session.commit()
            logger.info(f"Comment added to task {task_id} by user {author_id}")
            return comment
        except Exception as e:
            db.session.rollback()
            logger.error(f"Error adding comment to task {task_id}: {e}")
            raise

    @staticmethod
    @log_service_operation
    def get_task_comments(task_id, page=1, per_page=10):
        task = TaskService.get_task_by_id(task_id)
        comments = Comment.query.filter_by(task_id=task.id).order_by(Comment.created_at.desc()).paginate(page=page, per_page=per_page, error_out=False)
        return comments

    @staticmethod
    @log_service_operation
    def get_comment_by_id(comment_id):
        comment = Comment.query.get(comment_id)
        if not comment:
            raise ResourceNotFound(f"Comment with ID '{comment_id}' not found.")
        return comment

    @staticmethod
    @log_service_operation
    def update_comment(comment_id, data):
        comment = TaskService.get_comment_by_id(comment_id)
        try:
            comment.content = data.get('content', comment.content)
            db.session.commit()
            logger.info(f"Comment {comment_id} updated.")
            return comment
        except Exception as e:
            db.session.rollback()
            logger.error(f"Error updating comment {comment_id}: {e}")
            raise

    @staticmethod
    @log_service_operation
    def delete_comment(comment_id):
        comment = TaskService.get_comment_by_id(comment_id)
        try:
            db.session.delete(comment)
            db.session.commit()
            logger.info(f"Comment {comment_id} deleted.")
            return True
        except Exception as e:
            db.session.rollback()
            logger.error(f"Error deleting comment {comment_id}: {e}")
            raise

```

File: project-{topic}-complete/src/services/TaskService.cpp
```cpp
#include "TaskService.h"
#include "db/Database.h"
#include "utils/Logger.h"
#include <sstream>
#include <algorithm> // For std::remove_if

// Initialize static members for caching
std::map<long, Task> TaskService::s_task_cache;
std::mutex TaskService::s_cache_mutex;
const std::chrono::seconds TaskService::s_cache_ttl = std::chrono::minutes(5); // Cache for 5 minutes

Task TaskService::createTask(const Task& newTask) {
    if (newTask.title.empty() || newTask.userId == 0) {
        throw std::runtime_error("Task title and user ID cannot be empty.");
    }

    std::string sql = "INSERT INTO tasks (user_id, title, description, status, due_date) VALUES ($1, $2, $3, $4, $5) RETURNING id, user_id, title, description, status, due_date, created_at, updated_at;";
    std::vector<std::string> params = {
        std::to_string(newTask.userId),
        newTask.title,
        newTask.description.has_value() ? newTask.description.value() : "",
        newTask.status,
        newTask.dueDate.has_value() ? newTask.dueDate.value() : ""
    };

    try {
        std::vector<pqxx::row> rows = Database::executeQuery(sql, params);
        if (rows.empty()) {
            throw std::runtime_error("Failed to create task, no data returned.");
        }
        const pqxx::row& row = rows[0];
        Task createdTask;
        createdTask.id = row["id"].as<long>();
        createdTask.userId = row["user_id"].as<long>();
        createdTask.title = row["title"].as<std::string>();
        createdTask.description = row["description"].is_null() ? std::nullopt : std::make_optional(row["description"].as<std::string>());
        createdTask.status = row["status"].as<std::string>();
        createdTask.dueDate = row["due_date"].is_null() ? std::nullopt : std::make_optional(row["due_date"].as<std::string>());
        createdTask.createdAt = row["created_at"].as<std::string>();
        createdTask.updatedAt = row["updated_at"].as<std::string>();

        addToCache(createdTask); // Add to cache
        LOG_INFO("Task created: ID={}, UserID={}, Title='{}'", createdTask.id, createdTask.userId, createdTask.title);
        return createdTask;
    } catch (const DbException& e) {
        LOG_ERROR("Database error creating task: {}", e.what());
        throw std::runtime_error("Database error creating task.");
    }
}

std::optional<Task> TaskService::getTaskById(long taskId, long userId) {
    // Try to get from cache first
    std::lock_guard<std::mutex> lock(s_cache_mutex);
    if (s_task_cache.count(taskId)) {
        Task cachedTask = s_task_cache[taskId];
        // Basic cache validation: check user ID and TTL (if a timestamp was stored)
        if (cachedTask.userId == userId) { // && (now - cachedTask.timestamp) < s_cache_ttl
            LOG_DEBUG("Task {} retrieved from cache.", taskId);
            return cachedTask;
        } else {
            // Task in cache belongs to different user or is invalid for this request
            removeFromCache(taskId);
        }
    }
    lock.unlock(); // Release lock before going to DB

    std::string sql = "SELECT id, user_id, title, description, status, due_date, created_at, updated_at FROM tasks WHERE id = $1 AND user_id = $2;";
    std::vector<std::string> params = {std::to_string(taskId), std::to_string(userId)};

    try {
        std::vector<pqxx::row> rows = Database::executeQuery(sql, params);
        if (rows.empty()) {
            return std::nullopt;
        }
        const pqxx::row& row = rows[0];
        Task task;
        task.id = row["id"].as<long>();
        task.userId = row["user_id"].as<long>();
        task.title = row["title"].as<std::string>();
        task.description = row["description"].is_null() ? std::nullopt : std::make_optional(row["description"].as<std::string>());
        task.status = row["status"].as<std::string>();
        task.dueDate = row["due_date"].is_null() ? std::nullopt : std::make_optional(row["due_date"].as<std::string>());
        task.createdAt = row["created_at"].as<std::string>();
        task.updatedAt = row["updated_at"].as<std::string>();

        addToCache(task); // Add to cache after retrieval
        LOG_DEBUG("Task {} retrieved from DB.", taskId);
        return task;
    } catch (const DbException& e) {
        LOG_ERROR("Database error getting task by ID {}: {}", taskId, e.what());
        throw std::runtime_error("Database error retrieving task.");
    }
}

std::vector<Task> TaskService::getTasksByUserId(long userId) {
    std::vector<Task> tasks;
    std::string sql = "SELECT id, user_id, title, description, status, due_date, created_at, updated_at FROM tasks WHERE user_id = $1 ORDER BY created_at DESC;";
    std::vector<std::string> params = {std::to_string(userId)};

    try {
        std::vector<pqxx::row> rows = Database::executeQuery(sql, params);
        for (const auto& row : rows) {
            Task task;
            task.id = row["id"].as<long>();
            task.userId = row["user_id"].as<long>();
            task.title = row["title"].as<std::string>();
            task.description = row["description"].is_null() ? std::nullopt : std::make_optional(row["description"].as<std::string>());
            task.status = row["status"].as<std::string>();
            task.dueDate = row["due_date"].is_null() ? std::nullopt : std::make_optional(row["due_date"].as<std::string>());
            task.createdAt = row["created_at"].as<std::string>();
            task.updatedAt = row["updated_at"].as<std::string>();
            tasks.push_back(task);
        }
        LOG_DEBUG("Retrieved {} tasks for user ID {}.", tasks.size(), userId);
        return tasks;
    } catch (const DbException& e) {
        LOG_ERROR("Database error getting tasks for user ID {}: {}", userId, e.what());
        throw std::runtime_error("Database error retrieving tasks.");
    }
}

Task TaskService::updateTask(long taskId, long userId, const Task& updatedTaskData) {
    std::stringstream ss;
    ss << "UPDATE tasks SET ";
    std::vector<std::string> params;
    int param_idx = 1;

    // Dynamically build update query based on provided fields
    if (!updatedTaskData.title.empty()) {
        ss << "title = $" << param_idx++ << ", ";
        params.push_back(updatedTaskData.title);
    }
    if (updatedTaskData.description.has_value()) {
        ss << "description = $" << param_idx++ << ", ";
        params.push_back(updatedTaskData.description.value());
    } else if (updatedTaskData.description.has_value() == false && updatedTaskData.description.value() == "") {
        // If an empty string is explicitly provided for description, set it to NULL
        ss << "description = NULL, ";
    }

    // Validate status against enum values
    if (!updatedTaskData.status.empty()) {
        if (updatedTaskData.status != "TODO" && updatedTaskData.status != "IN_PROGRESS" && updatedTaskData.status != "DONE") {
            throw std::runtime_error("Invalid task status: " + updatedTaskData.status);
        }
        ss << "status = $" << param_idx++ << ", ";
        params.push_back(updatedTaskData.status);
    }
    if (updatedTaskData.dueDate.has_value()) {
        ss << "due_date = $" << param_idx++ << ", ";
        params.push_back(updatedTaskData.dueDate.value());
    } else if (updatedTaskData.dueDate.has_value() == false && updatedTaskData.dueDate.value() == "") {
        // If an empty string is explicitly provided for due_date, set it to NULL
        ss << "due_date = NULL, ";
    }

    ss << "updated_at = CURRENT_TIMESTAMP WHERE id = $" << param_idx++ << " AND user_id = $" << param_idx++ << " RETURNING id, user_id, title, description, status, due_date, created_at, updated_at;";
    
    // Remove trailing ", " if any
    std::string sql = ss.str();
    if (sql.find(", updated_at") != std::string::npos) { // Check if there was at least one field updated before updated_at
        sql.replace(sql.find("updated_at") - 2, 2, ""); // Remove ", "
    }

    params.push_back(std::to_string(taskId));
    params.push_back(std::to_string(userId));

    try {
        std::vector<pqxx::row> rows = Database::executeQuery(sql, params);
        if (rows.empty()) {
            throw std::runtime_error("Task not found or unauthorized to update.");
        }
        const pqxx::row& row = rows[0];
        Task updatedTask;
        updatedTask.id = row["id"].as<long>();
        updatedTask.userId = row["user_id"].as<long>();
        updatedTask.title = row["title"].as<std::string>();
        updatedTask.description = row["description"].is_null() ? std::nullopt : std::make_optional(row["description"].as<std::string>());
        updatedTask.status = row["status"].as<std::string>();
        updatedTask.dueDate = row["due_date"].is_null() ? std::nullopt : std::make_optional(row["due_date"].as<std::string>());
        updatedTask.createdAt = row["created_at"].as<std::string>();
        updatedTask.updatedAt = row["updated_at"].as<std::string>();

        updateCache(updatedTask); // Update cache
        LOG_INFO("Task updated: ID={}, UserID={}, Title='{}'", updatedTask.id, updatedTask.userId, updatedTask.title);
        return updatedTask;
    } catch (const DbException& e) {
        LOG_ERROR("Database error updating task {}: {}", taskId, e.what());
        throw std::runtime_error("Database error updating task.");
    }
}

bool TaskService::deleteTask(long taskId, long userId) {
    std::string sql = "DELETE FROM tasks WHERE id = $1 AND user_id = $2;";
    std::vector<std::string> params = {std::to_string(taskId), std::to_string(userId)};

    try {
        int affected_rows = Database::executeCommand(sql, params);
        if (affected_rows > 0) {
            removeFromCache(taskId); // Remove from cache
            LOG_INFO("Task deleted: ID={}, UserID={}", taskId, userId);
            return true;
        }
        LOG_WARN("Task not found or unauthorized to delete: ID={}, UserID={}", taskId, userId);
        return false;
    } catch (const DbException& e) {
        LOG_ERROR("Database error deleting task {}: {}", taskId, e.what());
        throw std::runtime_error("Database error deleting task.");
    }
}

void TaskService::addToCache(const Task& task) {
    std::lock_guard<std::mutex> lock(s_cache_mutex);
    s_task_cache[task.id] = task;
    LOG_DEBUG("Task {} added to cache.", task.id);
}

void TaskService::updateCache(const Task& task) {
    std::lock_guard<std::mutex> lock(s_cache_mutex);
    if (s_task_cache.count(task.id)) {
        s_task_cache[task.id] = task; // Overwrite with updated data
        LOG_DEBUG("Task {} updated in cache.", task.id);
    } else {
        s_task_cache[task.id] = task; // Add if not present
        LOG_DEBUG("Task {} added to cache during update (was not found).", task.id);
    }
}

void TaskService::removeFromCache(long taskId) {
    std::lock_guard<std::mutex> lock(s_cache_mutex);
    if (s_task_cache.count(taskId)) {
        s_task_cache.erase(taskId);
        LOG_DEBUG("Task {} removed from cache.", taskId);
    }
}

void TaskService::clearCacheForUser(long userId) {
    // For a map keyed by task ID, clearing for a specific user requires iterating.
    // In a production cache, you might have user-specific caches or a more complex eviction policy.
    std::lock_guard<std::mutex> lock(s_cache_mutex);
    auto it = s_task_cache.begin();
    while (it != s_task_cache.end()) {
        if (it->second.userId == userId) {
            it = s_task_cache.erase(it);
        } else {
            ++it;
        }
    }
    LOG_DEBUG("Cache cleared for tasks belonging to user {}.", userId);
}
```